Add ability to change background dim colour in grayscale

Uncomfy opened the pull request:
Previously discussed here: https://github.com/ppy/osu/discussions/12983

Depends on:
- https://github.com/ppy/osu-resources/pull/344
- https://github.com/ppy/osu-framework/pull/6395
- https://github.com/ppy/osu-framework/pull/6404
- https://github.com/ppy/osu-framework/pull/6591

Allows to change which colour the background is dimmed to. Currently only supports grayscale.

This is not a full implementation, as it doesn't support storyboards and videos.

The main reason I'm asking for comments so early is that this implementation handles dimming in two different classes, which leads to code duplication, so I wonder if there is a cleaner way to do this.

https://github.com/user-attachments/assets/5be01be4-7bae-4d9d-b8cb-057d933abb0e

Here are also a few examples of how it could look with colour picker instead of a grayscale slider, using [SettingsColour](https://github.com/ppy/osu/blob/71eb7120bad1b6e6ab778c13695fc568becaf3c5/osu.Game/Overlays/Settings/SettingsColour.cs), but I'm unsure if it fits the design, as colour pickers weren't used in Settings/Visual Settings before. Also it doesn't work properly in Visual Settings yet, because it doesn't have a PopoverContainer in its hierarchy, and I'm not really sure how to handle it.

![settings](https://github.com/user-attachments/assets/f4eb0f3d-f816-487c-a9db-d260b7ccff81)
![settings_colour_picker_open](https://github.com/user-attachments/assets/779bf184-32a3-456c-9aef-f7ccbd63a522)
![visual_settings_1](https://github.com/user-attachments/assets/b314c9d9-5409-4382-8c64-6e8cc643f785)

Uncomfy committed:
Added "Background Colour" setting

Uncomfy committed:
Made Background.Sprite replaceable

Uncomfy committed:
Made a custom class for BeatmapBackground with a custom BeatmapBackgroundSprite

Uncomfy committed:
Made BeatmapBackgroundSprite use a custom shader

Uncomfy committed:
Added a custom DrawNode for BeatmapBackgroundSprite

Uncomfy committed:
Passing parameters to the shader

Uncomfy committed:
Made BeatmapBackground shader handle dimming

Uncomfy committed:
Added support for DimColour

Uncomfy committed:
Pass a full colour for DimColour instead of a single value into a shader

Uncomfy committed:
Changed BeatmapBackground class name to DimmableBeatmapBackground to avoid confusion with existing BeatmapBackground class

Uncomfy committed:
Moved DimmableBeatmapBackground to its own file

Uncomfy committed:
Made DimmableBeatmapBackground inherit from BeatmapBackground, switched BeatmapBackground to DimmableBeatmapBackground in BackgroundScreenBeatmap

Uncomfy committed:
Switched to using Color4 for DimColour instead of float

Uncomfy committed:
Made DimmableBeatmapBackgroundSprite invalidate DrawNode on DimColour and DimLevel update

Uncomfy committed:
Formatting fixes

Uncomfy committed:
Changed the way colour is reported in TestSceneUserDimBackgrounds.cs

Uncomfy committed:
Added DimColour to user settings in TestSceneUserDimBackgrounds.cs

Uncomfy committed:
Updated TestSceneUserDimBackgrounds.cs to test for colour offset caused by DimColour

Uncomfy committed:
Handle IgnoreUserSettings for DimColour

Uncomfy committed:
Updated the way TestSceneUserDimBackgrounds.cs checks if background is black

Uncomfy committed:
Integrated DimmableBeatmapBackground into BeatmapBackground

Uncomfy committed:
Use existing methods to update shader variables

Uncomfy committed:
Made background receive correct dim values when it's set

Uncomfy committed:
Disable DimColour when user settings are ignored

Uncomfy committed:
Made BufferedContainer handle dimming when it's present

Uncomfy committed:
Added default values for BeatmapBackgroundSprite

Uncomfy committed:
Added explanations on why and how colour is split into two components

Uncomfy committed:
Updated the way colours are computed and checked in TestSceneUserDimBackgrounds

Uncomfy committed:
Added forgotten Dispose's

Uncomfy committed:
Moved BeatmapBackground creation into a virtual function to help with testing

Uncomfy committed:
Created TestBeatmapBackground class and moved colour computations into it

Uncomfy committed:
Added an interface to simplify access to current dimmable object in BeatmapBackground

Uncomfy committed:
Formatting fixes

Uncomfy committed:
Get DrawColour directly from object drawing the BG in TestSceneUserDimBackgrounds

Uncomfy committed:
Changed the way Background.Sprite is replaced by BeatmapBackground

Uncomfy committed:
Changed the way Background.bufferedContainer is replaced by BeatmapBackground

Uncomfy committed:
Added test to verify that both dimming handlers work as intended

Uncomfy committed:
Made TestSceneUserDimBackgrounds compute current colour using ColoredDimmable's DimLevel and DimColour

Uncomfy committed:
Added Background colour to visual settings

Uncomfy committed:
Changed BackgroundColour localisable string into BackgroundDimColour

Uncomfy committed:
Changed the way ParentDrawColour is computed

Uncomfy committed:
Updated the source of DrawColour when computing expected/target colour

Uncomfy committed:
Formatting fixes

Uncomfy committed:
Small formatting fixes

Uncomfy committed:
Removed dependency on osuTK

Uncomfy committed:
Preload BeatmapBackground shader

jhk2601 commented:
Just a quick thing, "0%-100%" doesn't really make sense in the context of a slider changing how light or dark the dim is. This would be fixed in the color picker implementation anyway though (which I personally think makes more sense, though maybe hiding the UI element behind an option that's off by default?)

frenzibyte commented on osu.Game/Graphics/Backgrounds/BeatmapBackground.cs:
```diff
@@ -35,5 +96,209 @@ public override bool Equals(Background other)
+            [BackgroundDependencyLoader]
+            private void load(ShaderManager shaders)
+            {
+                TextureShader = shaders.Load(VertexShaderDescriptor.TEXTURE_2, "BeatmapBackground");
```
Isn't `TextureShader` responsible for rendering the framebuffer texture of the buffered container? If so, isn't that pointless? Isn't it enough to apply the custom "BeatmapBackground" shader on the underlying beatmap background sprite? Would that cause issues with blurring? I can't imagine it will.

Uncomfy replied:
Yes, but it will require the redraw of the framebuffer texture on each DimLevel change, which will also force the blur to be redrawn, and I feared that this might introduce unnecessary overhead during (for example) breaks, if player has background blur enabled. If that is not a concern - yeah, keeping the custom shader only on the sprite makes complete sense.

Uncomfy replied:
You're right, it's pointless, at least in the current implementation.
Changing the dim parameters requires DrawNode invalidation, which forces the framebuffer to be redrawn, so there are no benefits. This does mean that DimLevel transitions are way more expensive now (at least on my rather old and slow hardware).

Here are the results of my benchmarks for average frametime during different parts of the beatmap.
With user blur set to 0%:
| |Gameplay|Breaks|Fade in|Fade out|
|---|---|---|---|---|
|Baseline|4.25 ms|4.29 ms|4.42 ms|4.08 ms|
|Coloured dim|4.26 ms|4.33 ms|4.92 ms|4.55 ms|
|Relative increase|0.2%|0.9%|**11.3%**|**11.5%**|

With user blur set to 100%:
| |Gameplay|Breaks|Fade in|Fade out|
|---|---|---|---|---|
|Baseline|4.20 ms|4.24 ms|4.39 ms|4.02 ms|
|Coloured dim|4.22 ms|4.30 ms|5.18 ms|4.66 ms|
|Relative increase|0.5%|1.4%|**18%**|**15.9%**|

"Baseline" is osu! 2024.1009.1
"Coloured dim" is commit [c5867b6](https://github.com/ppy/osu/pull/30391/commits/c5867b6dd9db22935e574cb2d8e7755125ece597)

<details>
<summary>Details about the benchmark</summary>

Made a map with 64 gameplay sections and 63 breaks for this benchmark, total duration is 3:09.

"Fade in" refers to frames rendered during gameplay-to-break transition, "Fade out" refers to break-to-gameplay transition.
"Gameplay" and "Breaks" columns are just for sanity-checking, as they should be same.

Total amount of recorded frames:
| |Blur|Gameplay|Breaks|Fade in|Fade out|
|---|---|---|---|---|---|
|Baseline|0%|12068|8176|11018|12069|
|Baseline|100%|12193|8253|11089|12236|
|Coloured dim|0%|12090|8105|9810|10689|
|Coloured dim|100%|12154|8137|9448|10512|

osu! settings:
- Compiled in release mode
- Single-threaded mode
- "OpenGL (Experimental)" renderer
- Windowed screen mode
- Resolution: 1564x964
- Frame limiter: Basically unlimited

System info:
- OS: Kubuntu 24.04.1 LTS x86_64
- CPU: Intel i3-6100U @ 2.300GHz
- GPU: Intel HD Graphics 520
- RAM: 19907MiB

</details>

So I guess now there are two ways to go - ditch the shader replacement on the BufferedContainer and have slower transitions, or modify the BufferedContainer somehow to avoid redrawing the framebuffer when shader parameters are changed.

Which one is more preferable?

frenzibyte replied:
Fair enough, this is a bit of a pesky situation. You can control whether the framebuffer gets redrawn by overriding `GetDrawVersion()` and make that point to a custom invalidation ID that is normally incremented but not when dim properties are changed. See `Path` in osu!framework, it might actually be doing exactly what you're looking for.

Uncomfy replied:
Added a test case for it, but solving it requires a change in the framework: https://github.com/ppy/osu-framework/pull/6404

With that change applied benchmarks look a lot better.

With user blur set to 0%:
|Implementation|Gameplay|Breaks|Fade in|Fade out|
|---|---|---|---|---|
|Baseline|4.15 ms|4.20 ms|4.34 ms|3.99 ms|
|Coloured dim|4.19 ms|4.23 ms|4.39 ms|4.00 ms|
|Relative increase|1.0%|0.7%|1.2%|0.3%|

With user blur set to 100%:
|Implementation|Gameplay|Breaks|Fade in|Fade out|
|---|---|---|---|---|
|Baseline|4.11 ms|4.16 ms|4.31 ms|3.93 ms|
|Coloured dim|4.05 ms|4.11 ms|4.25 ms|3.85 ms|
|Relative increase|-1.5%|-1.2%|-1.4%|-2.0%|

"Baseline" is osu! 2024.1009.1
"Coloured dim" is commit https://github.com/ppy/osu/pull/30391/commits/93c3fac26e55f90f34ff625afa0aec0e4e9f9080 with https://github.com/ppy/osu-framework/pull/6404 applied

peppy replied:
I don't think these benchmark numbers represent the true impact because as far as I read you're measuring over a long period of time, while when the blur re-renders it's going to be very detrimental only to the local frames (so the overall "increase" is much smaller than the actual perceived difference to the user).

Of course, as long as the numbers are going down this doesn't matter so much, so carry on I guess.

A better way of benchmarking would be to force the redraw every frame then just take a reading of the average frame time.

Uncomfy replied:
I've double checked, and without https://github.com/ppy/osu-framework/pull/6404 my implementation triggered a blur redraw on every frame during transition, but didn't affect frames during gameplay or break.
This means that the stats computed in the "Fade in" and "Fade out" columns only take into account frames where redraw happened.

> A better way of benchmarking would be to force the redraw every frame then just take a reading of the average frame time.
My goal was to show that extra redraws were triggered and that it affected the performance during specific parts of the beatmap, so I couldn't use this approach

Uncomfy commented:
> Just a quick thing, "0%-100%" doesn't really make sense in the context of a slider changing how light or dark the dim is. This would be fixed in the color picker implementation anyway though (which I personally think makes more sense, though maybe hiding the UI element behind an option that's off by default?)
I'm not sure what would be better than 0%-100% for a slider. Or do you mean ditching the slider completely?
Would renaming the setting to background dim brightness make more sense?

jhk2601 commented:
> > Just a quick thing, "0%-100%" doesn't really make sense in the context of a slider changing how light or dark the dim is. This would be fixed in the color picker implementation anyway though (which I personally think makes more sense, though maybe hiding the UI element behind an option that's off by default?)
> 
> I'm not sure what would be better than 0%-100% for a slider. Or do you mean ditching the slider completely? Would renaming the setting to background dim brightness make more sense?

yes that would be better, if I saw a slider labeled "background dim color" and it was a percent I wouldn't know what the number represented, "brightness" works better.

peppy commented:
It might make more sense to rename "Background dim" to "Background opacity" and have 100% be a fully visible image. Although this will go against user expectations coming from stable so maybe it's too late to make a change like this.

Uncomfy committed:
Test if BufferedContainer's framebuffer is redrawn on dim changes

Uncomfy commented:
I wonder if it is fine to just add an overlay (a quad with `Colour == DimColour`, `Alpha == DimLevel` and a size that covers an entire background) to achieve same effect for storyboards/videos.
It's a way simpler solution than changing shaders of all of the `osu.Game.Storyboards.Drawables`. It might cause performance to drop a bit, though I'm unsure how noticeable it would be with storyboards/videos enabled.

frenzibyte commented:
I can only understand doing this for storyboards since they are not buffered (and I would question how it's supposed to behave, because blending the entire storyboard behind some "background dim colour" overlay might end up looking silly). For videos, this can still totally be done by shaders with some changes framework-side.

frenzibyte commented:
"Just adding an overlay quad with..." is exactly the 1x overhead problem that we would like to avoid here by doing better.

Uncomfy commented:
Yeah, true
I just wondered if it's going to be fine at least with storyboards, given that they're pretty heavy already (though it depends on individual storyboards), so the performance hit might not be as noticeable
But ok, no shortcuts

Uncomfy committed:
Replace the shader of the DrawableStoryboardSprite

Uncomfy committed:
Replace the shader of the DrawableStoryboardAnimation

Uncomfy committed:
Changed the name of the shader

Uncomfy committed:
Replace the shader of the DrawableStoryboardVideo

Uncomfy committed:
Update dim shader's parameters

Uncomfy committed:
Removed direct reliance on DimLevel for coloured dim shaders

Uncomfy committed:
Fix exception message

Uncomfy committed:
Made regular background use DrawColourOffset

...for changing the dim colour.
It is computed in a similar way to Drawable.DrawColourInfo

Uncomfy committed:
Support for coloured dimming in storyboards and videos

Uncomfy committed:
Use default implementation in interface to reduce code duplication

Uncomfy committed:
Fixed DrawColourOffset being applied twice to DimmableSprite in DimmableBufferedContainer

Uncomfy committed:
Disable DrawColourOffset on sprites with additive blending

Need to double-check that there is always going to be something beneath them that will have the offset.

Uncomfy committed:
Remove unused "using" directives

Uncomfy committed:
Made tests compile

Uncomfy committed:
Fixed IsBackgroundDimmed and IsBackgroundColourOffset

Uncomfy committed:
Fix BufferedContainer dim check

Uncomfy committed:
Make only Invalidation.Colour invalidate coloured dimmable DrawNodes

Uncomfy committed:
Disable Invalidation.Colour propagation in DimmableBufferedContainer

This stops the framebuffer from being redrawn when colour changes
...which only happens because ColouredDimmableSprite needs to invalidate
DrawNode together with the colour change.

Uncomfy committed:
Reduce code duplication by reusing DimmableSprite

Uncomfy committed:
Added test to check if storyboard elements receive correct DrawColourOffset

Uncomfy committed:
Spellfixes

Uncomfy committed:
Merge tag '2025.607.0' into background_dim_colour

Uncomfy committed:
Return accidentally removed copyright comment, oops

Uncomfy committed:
Set BeatmapBackground as default Background in DimmableBackground

Uncomfy committed:
Remove "?" from variables since nullable is disabled

Uncomfy committed:
Move IColouredDimmable interface into its own file

Uncomfy committed:
Moved DimmableSprite to its own file

Uncomfy committed:
Moved DimmableBufferedContainer to its own file

Uncomfy committed:
Naming changes

Uncomfy committed:
Fix InspectCode complaints

Uncomfy committed:
Made ColouredDimmableBufferedContainer invalidate DrawNode on Colour invalidation before removing Colour from invalidation

Uncomfy committed:
Comment clarification

Uncomfy commented:
Storyboards are now also supported

https://github.com/user-attachments/assets/89e11fdc-0980-4b3b-a5a2-6a744918c30f

However, I used default method implementation in `IColouredDimmable` interface, unsure if that's ok to do

Uncomfy commented:
I got stuck here and don't have much time lately to figure out a solution.

Performance degradation is caused by the need to update shader parameters before drawing each storyboard element, which introduces a huge overhead. Someone (whoever it was - sorry for forgetting you name) suggested updating the shader params once before storyboard is drawn. That wasn't as straightforward, because custom shader has different behavior for elements that have additive and regular (`color * alpha + dest * (1 - alpha)`) blending. Additive elements should ignore the custom background color because elements underneath already have it added to them.

Now that I've written this comment - I wonder if it's possible to standardize shader behavior for all elements.
